src/toast/_libtoast/accelerator.cpp

```diff
@@ -0,0 +1,401 @@
+
+            bool present = acc_is_present(p_host, n_bytes);
+            if (!present) {
+                std::ostringstream o;
+                o << "Data is not present on device, cannot update host.";
+                log.error(o.str().c_str());
+                throw std::runtime_error(
+                          o.str().c_str());
+            }
+
+            std::ostringstream o;
+            o << "update host/self with host pointer " << p_host << " (" << n_bytes << " bytes)";
+            log.verbose(o.str().c_str());
+
+            # ifdef USE_OPENACC_MEMPOOL
+
+            // FIXME: Does the pool have an update device / host equivalent?
+            auto & pool = GPU_memory_pool::get();
+            pool.fromDevice(reinterpret_cast <char *> (p_host));
+            # else // ifdef USE_OPENACC_MEMPOOL
+            acc_update_self(p_host, n_bytes);
+            # endif // ifdef USE_OPENACC_MEMPOOL
+            return;
+        }, py::arg(
+            "data"), R"(
+        Update host copy of the data from the device.
+
+        Args:
+            data (array):  The host data.
+
+        Returns:
+            None
+
+    )");
+
+    m.def(
+        "acc_delete", [](py::buffer data) {
+            auto & log = toast::Logger::get();
+            py::buffer_info info = data.request();
+            void * p_host;
+            size_t n_elem;
+            size_t n_bytes;
+            extract_buffer_info(info, &p_host, &n_elem, &n_bytes);
+
+            bool present = acc_is_present(p_host, n_bytes);
+            if (!present) {
+                std::ostringstream o;
+                o << "Data is not present on device, cannot delete.";
+                log.error(o.str().c_str());
+                throw std::runtime_error(
+                          o.str().c_str());
+            }
+
+            std::ostringstream o;
+            o << "delete device mem for host pointer " << p_host << " (" << n_bytes << " bytes)";
+            log.verbose(o.str().c_str());
+
+            # ifdef USE_OPENACC_MEMPOOL
+
+            // FIXME:  If there was a pool.Free(cpu pointer), we would call that here.
+            auto & pool = GPU_memory_pool::get();
+            pool.fromDevice(reinterpret_cast <char *> (p_host));
+            # else // ifdef USE_OPENACC_MEMPOOL
+            acc_delete(p_host, n_bytes);
+            # endif // ifdef USE_OPENACC_MEMPOOL
+            return;
+        }, py::arg(
+            "data"), R"(
+        Delete the device copy of the data.
+
+        Args:
+            data (array):  The host data.
+
+        Returns:
+            None
+
+    )");
+
+    #else // ifdef HAVE_OPENACC
+
+    register_stub(m, "acc_is_present");
+    register_stub(m, "acc_copyin");
+    register_stub(m, "acc_copyout");
+    register_stub(m, "acc_update_device");
+    register_stub(m, "acc_update_self");
+    register_stub(m, "acc_delete");
+
+    #endif // HAVE_OPENACC
+
+    // Small test code used by the unit tests.
```

Comment:
This is a curious place to add test code. Is this the permanent location?

Reply:
It is compiled code that is only used in the unit tests for this source file.  I suppose we should eventually put it in a separate file like `_libtoast/test_accelerator.cpp`.